File: src/translator/zh2en/provider/skill.ts
import { DATA as POE_DATA } from "../../../data/poe/index.js";
import type { Skill } from "../../../data/poe/types.js";

export class SkillProvider {
    private readonly zhIdx = new Map<string, Skill>();
    private readonly indexableSupportZhIdx = new Map<string, Skill>();

    constructor() {
        for (const skill of POE_DATA.gemSkills) {
            this.zhIdx.set(skill.zh, skill);
        }
        for (const skill of POE_DATA.hybridSkills) {
            this.zhIdx.set(skill.zh, skill);
        }
        for (const skill of POE_DATA.transfiguredSkills) {
            this.zhIdx.set(skill.zh, skill);
        }

        for (const skill of POE_DATA.indexableSupports) {
            this.indexableSupportZhIdx.set(skill.zh, skill);
        }
    }

    provideSkill(zh: string): Skill | undefined {
        return this.zhIdx.get(zh);
    }

    provideIndexableSupport(zh: string): Skill | undefined {
        return this.indexableSupportZhIdx.get(zh);
    }
}

File: go/translator/zh2en/provider/skill.go
package provider

import (
	"github.com/cn-poe-community/cn-poe-utils/go/data/poe"
)

type SkillProvider struct {
	zhIdx                 map[string]*poe.Skill
	indexableSupportZhIdx map[string]*poe.Skill
}

func NewSkillProvider() *SkillProvider {
	zhIdx := make(map[string]*poe.Skill)
	indexableSupportZhIdx := make(map[string]*poe.Skill)

	for i := range poe.DATA.GemSkills {
		skill := &poe.DATA.GemSkills[i]
		zhIdx[skill.Zh] = skill
	}
	for i := range poe.DATA.HybridSkills {
		skill := &poe.DATA.HybridSkills[i]
		zhIdx[skill.Zh] = skill
	}
	for i := range poe.DATA.TransfiguredSkills {
		skill := &poe.DATA.TransfiguredSkills[i]
		zhIdx[skill.Zh] = skill
	}

	for i := range poe.DATA.IndexableSupports {
		skill := &poe.DATA.IndexableSupports[i]
		indexableSupportZhIdx[skill.Zh] = skill
	}
	return &SkillProvider{
		zhIdx:                 zhIdx,
		indexableSupportZhIdx: indexableSupportZhIdx,
	}
}

func (p *SkillProvider) ProvideSkill(name string) *poe.Skill {
	if skill, ok := p.zhIdx[name]; ok {
		return skill
	}
	return nil
}

func (p *SkillProvider) ProvideIndexableSupport(name string) *poe.Skill {
	if skill, ok := p.indexableSupportZhIdx[name]; ok {
		return skill
	}
	return nil
}
